storage/innobase/row/row0ins.cc

```diff
@@ -2708,33 +2708,28 @@ row_ins_clust_index_entry_low(
 	    && !entry->is_metadata() && !trx->duplicates
 	    && !trx->check_unique_secondary && !trx->check_foreigns
 	    && !trx->dict_operation
-	    && block->page.id().page_no() == index->page
-	    && !index->table->skip_alter_undo
-	    && !index->table->n_rec_locks
-	    && !index->table->is_active_ddl()
-	    && !index->table->has_spatial_index()
-	    && !index->table->versioned()
-            && (!dict_table_is_partition(index->table)
+	    && (!dict_table_is_partition(index->table)
 	        || thd_sql_command(trx->mysql_thd) == SQLCOM_INSERT)) {
-		DEBUG_SYNC_C("empty_root_page_insert");
 
-		trx->bulk_insert = true;
-
-		if (!index->table->is_temporary()) {
+		if (!index->table->has_spatial_index()
+		    && !index->table->n_rec_locks
+		    && !index->table->versioned()
+		    && !index->table->skip_alter_undo) {
+			trx->bulk_insert= true;
+			if (index->table->is_temporary()) {
+				goto skip_bulk_insert;
+			}
 			err = lock_table(index->table, NULL, LOCK_X, thr);
```

Comment:
Can we avoid most of these changes? Especially the `goto skip_bulk_insert` after enabling bulk insert for temporary tables is rather confusing, although it is not logically incorrect.

I think that we should always have `index->table->skip_alter_undo==false` in this branch, so we’d better remove that condition and add `ut_ad(!index->table->skip_alter_undo)`.

Why are we removing the check that the current page must be the root page? This is the DML branch after all.

Can the condition on `!index->table->is_active_ddl()` actually ever hold here? If we are about to execute `ALTER TABLE…ALGORITHM=INPLACE, LOCK=NONE` on an empty table, don’t we handle that specially nowadays? I can only imagine that purge might be able to remove any history and make the table empty after `ha_innobase::prepare_inplace_alter_table()` returned. So, maybe that condition can actually hold and should not be removed.